go/pkg/services/logcollector/statefile.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2018 Iguazio
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logcollector
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/mlrun/mlrun/pkg/common"
+
+	"github.com/nuclio/errors"
+	"github.com/nuclio/logger"
+)
+
+type FileStateStore struct {
+	state                   *State
+	logger                  logger.Logger
+	stateFilePath           string
+	lock                    sync.Locker
+	stateFileUpdateInterval time.Duration
+}
+
+func NewStateFile(logger logger.Logger, filePath string, stateFileUpdateInterval time.Duration) *FileStateStore {
+	return &FileStateStore{
+		state: &State{
+			InProgress: make(map[string]LogItem),
+		},
+		logger:                  logger.GetChild("filestatestore"),
+		stateFilePath:           filePath,
+		stateFileUpdateInterval: stateFileUpdateInterval,
+		lock:                    &sync.Mutex{},
+	}
+}
+
+func (f *FileStateStore) AddLogItem(ctx context.Context, runId, selector string) error {
+	logItem := LogItem{
+		RunUID:        runId,
+		LabelSelector: selector,
+	}
+
+	if existingItem, exists := f.state.InProgress[runId]; exists {
+		f.logger.DebugWithCtx(ctx,
+			"Item already exists in state file. Overwriting label selector",
+			"runId", runId,
+			"existingItem", existingItem)
+	}
+
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	f.state.InProgress[logItem.RunUID] = logItem
+	return nil
+}
+
+func (f *FileStateStore) RemoveLogItem(runId string) error {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	delete(f.state.InProgress, runId)
+	return nil
+}
+
+func (f *FileStateStore) UpdateState(ctx context.Context) {
+
+	prevState := *f.state
+
+	for {
+
+		// get state
+		currentState := *f.state
+
+		// if state changed, write it to file
+		if !reflect.DeepEqual(currentState, prevState) {
+
+			prevState = currentState
+
+			// write state file
+			if err := f.writeStateToFile(ctx, &currentState); err != nil {
+				f.logger.ErrorWithCtx(ctx, "Failed to write state file", "err", err)
+				return
+			}
+		}
+
+		time.Sleep(f.stateFileUpdateInterval)
```

Comment:
always write to file, why do you care if has changed or not

Reply:
To minimize IO.
Why write to file if not changed?